File: solutions/Java/validate_binary_search_tree_inorder.java
public class Solution {
    long[] prev = {Long.MIN_VALUE};
    boolean dfs(TreeNode n) {
        if (n == null) return true;
        if (!dfs(n.left)) return false;
        if (n.val <= prev[0]) return false;
        prev[0] = n.val;
        return dfs(n.right);
    }
    public boolean isValidBST(TreeNode root) {
        return dfs(root);
    }
    public static class TreeNode { int val; TreeNode left, right; TreeNode(int x) { val = x; } }
}

File: solutions/Rust/validate_binary_search_tree_inorder.rs
use std::cell::RefCell; use std::rc::Rc;
#[derive(Debug)]
pub struct TreeNode { pub val: i32, pub left: Option<Rc<RefCell<TreeNode>>>, pub right: Option<Rc<RefCell<TreeNode>>>, }
pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
    let prev = std::cell::RefCell::new(i64::MIN);
    fn dfs(n: &Option<Rc<RefCell<TreeNode>>>, prev: &std::cell::RefCell<i64>) -> bool {
        if let Some(node) = n {
            let node_ref = node.borrow();
            if !dfs(&node_ref.left, prev) { return false; }
            let mut p = prev.borrow_mut();
            if (node_ref.val as i64) <= *p { return false; }
            *p = node_ref.val as i64;
            dfs(&node_ref.right, prev)
        } else { true }
    }
    dfs(&root, &prev)
}
